### PR TITLE
687 expose recent scheduling decisions

### PR DESCRIPTION
**Pull request by abhay-agarwal**
This creates a kv-store for data about declined offers for apps. If an
app’s resource requirements cannot be supplied by the current system,
there should eventually be some notification back to the executor.

Note: Manually adding event types to a bucket style method isn’t fun.

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/MarathonModule.scala [hunk 1]
```diff
@@ -163,6 +165,15 @@ class MarathonModule(conf: MarathonConf, http: HttpConf, zk: ZooKeeperClient)
 
   @Provides
   @Singleton
+  def provideTaskOffersDeclinedRepository(
+    state: State,
+    conf: MarathonConf,
+    registry: MetricRegistry): TaskOffersDeclinedRepository = {
```
Comment (drexin) [inline src/main/scala/mesosphere/marathon/MarathonModule.scala]: These parameters are unused, please remove them.

File: src/main/scala/mesosphere/marathon/MarathonScheduler.scala [hunk 2]
```diff
@@ -107,11 +107,22 @@ class MarathonScheduler @Inject() (
         log.debug("Received offer %s".format(offer))
 
         val matchingTask = taskQueue.pollMatching { app =>
-          newTask(app, offer).map(app -> _)
+          val result: TaskBuilder.BuildResult = newTask(app, offer)
+          result match {
+            case declined: TaskBuilder.BuildDeclined => {
```
Comment (drexin) [inline src/main/scala/mesosphere/marathon/MarathonScheduler.scala]: No need for curly braces here.

File: src/main/scala/mesosphere/marathon/MarathonScheduler.scala [hunk 3]
```diff
@@ -107,11 +107,22 @@ class MarathonScheduler @Inject() (
...
+              None
+            }
+            case success: TaskBuilder.BuildSuccess => Some((app, success))
+          }
         }
 
-        matchingTask.foreach {
-          case (app, (taskInfo, ports)) =>
+        matchingTask match {
+          case None => {
```
Comment (drexin) [inline src/main/scala/mesosphere/marathon/MarathonScheduler.scala]: No need for curly braces here.

File: src/main/scala/mesosphere/marathon/state/TaskOfferDeclinedRepository.scala [hunk 4]
```diff
@@ -0,0 +1,19 @@
+package mesosphere.marathon.state
+
+import mesosphere.marathon.event.TaskOfferDeclinedEvent
+
+import scala.collection.mutable
+
+class TaskOffersDeclinedRepository() {
+
+  protected[this] val tasksDeclined = mutable.Map[PathId, TaskOfferDeclinedEvent]()
```
Comment (drexin) [inline src/main/scala/mesosphere/marathon/state/TaskOfferDeclinedRepository.scala]: Please use TrieMap instead and remove the synchronized blocks.

File: src/main/scala/mesosphere/mesos/TaskBuilder.scala [hunk 5]
```diff
@@ -203,6 +202,10 @@ class TaskBuilder(app: AppDefinition,
 
 object TaskBuilder {
 
+  sealed trait BuildResult
+  case class BuildDeclined(reason: String) extends BuildResult
+  case class BuildSuccess(info: TaskInfo, ports: Seq[Long]) extends BuildResult
```
Comment (drexin) [inline src/main/scala/mesosphere/mesos/TaskBuilder.scala]: Please make both case classes `final`.

File: src/test/scala/mesosphere/mesos/TaskBuilderTest.scala [hunk 6]
```diff
@@ -149,9 +151,10 @@ class TaskBuilderTest extends MarathonSpec {
       )
     )
 
-    assert(task.isDefined)
+    assert(task.isInstanceOf[TaskBuilder.BuildSuccess])
 
-    val (taskInfo, taskPorts) = task.get
+    val success = task.asInstanceOf[TaskBuilder.BuildSuccess]
```
Comment (drexin) [inline src/test/scala/mesosphere/mesos/TaskBuilderTest.scala]: Please make this a `match`

File: src/test/scala/mesosphere/mesos/TaskBuilderTest.scala [hunk 7]
```diff
@@ -182,9 +185,10 @@ class TaskBuilderTest extends MarathonSpec {
...
```
Comment (drexin) [inline src/test/scala/mesosphere/mesos/TaskBuilderTest.scala]: Please make this a match

File: src/test/scala/mesosphere/mesos/TaskBuilderTest.scala [hunk 8]
```diff
@@ -214,9 +218,10 @@ class TaskBuilderTest extends MarathonSpec {
...
```
Comment (drexin) [inline src/test/scala/mesosphere/mesos/TaskBuilderTest.scala]: Please make this a `match`

File: src/test/scala/mesosphere/mesos/TaskBuilderTest.scala [hunk 9]
```diff
@@ -255,9 +260,10 @@ class TaskBuilderTest extends MarathonSpec {
...
```
Comment (drexin) [inline src/test/scala/mesosphere/mesos/TaskBuilderTest.scala]: Please make this a `match`

File: src/test/scala/mesosphere/mesos/TaskBuilderTest.scala [hunk 10]
```diff
@@ -326,20 +332,22 @@ class TaskBuilderTest extends MarathonSpec {
       s => TaskID(s.toString), taskTracker, defaultConfig())
 
     def shouldBuildTask(message: String, offer: Offer) {
-      val tupleOption = builder.buildIfMatches(offer)
-      assert(tupleOption.isDefined, message)
+      val task = builder.buildIfMatches(offer)
+      assert(task.isInstanceOf[TaskBuilder.BuildSuccess], message)
+
+      val success = task.asInstanceOf[TaskBuilder.BuildSuccess]
```
Comment (drexin) [inline src/test/scala/mesosphere/mesos/TaskBuilderTest.scala]: Please make this a `match`

File: src/test/scala/mesosphere/mesos/TaskBuilderTest.scala [hunk 11]
```diff
@@ -387,19 +395,22 @@ class TaskBuilderTest extends MarathonSpec {
...
```
Comment (drexin) [inline src/test/scala/mesosphere/mesos/TaskBuilderTest.scala]: Please make this a `match`

**Comment by drexin** [comment]
Thanks for your contribution. Please see my comments. Could you also please add an endpoint to retrieve the data and add it to the docs? And please squash all this into a single commit, thanks!

File: src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala [hunk 12]
```diff
@@ -457,10 +459,10 @@ class SchedulerActions(
       mapper
     )
 
-    builder.buildIfMatches(offer) map {
-      case (task, ports) =>
-        val taskBuilder = task.toBuilder
-        taskBuilder.build -> ports
+    builder.buildIfMatches(offer) match {
```
Comment (ConnorDoyle) [inline src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala]: It looks like this pattern match doesn't do anything.  In both cases, the result (or the equivalent) of the call to `buildIfMatches` is returned.  Why not just return that value instead?

File: src/main/scala/mesosphere/marathon/event/HistoryActor.scala [hunk 13]
```diff
@@ -20,6 +23,8 @@ class HistoryActor(eventBus: EventStream, taskFailureRepository: TaskFailureRepo
 
     case AppTerminatedEvent(appId, eventType, timestamp) =>
```
Comment (ConnorDoyle) [inline src/main/scala/mesosphere/marathon/event/HistoryActor.scala]: Upon receipt of an `AppTerminatedEvent`, the entry for the terminated app should be expunged also from the `taskOfferDeclinedRepository` here.

File: src/main/scala/mesosphere/mesos/TaskBuilder.scala [hunk 14]
```diff
@@ -31,7 +31,7 @@ class TaskBuilder(app: AppDefinition,
 
   val log = Logger.getLogger(getClass.getName)
 
-  def buildIfMatches(offer: Offer): Option[(TaskInfo, Seq[Long])] = {
+  def buildIfMatches(offer: Offer): TaskBuilder.BuildResult = {
```
Comment (ConnorDoyle) [inline src/main/scala/mesosphere/mesos/TaskBuilder.scala]: :100:

File: src/main/scala/mesosphere/mesos/TaskBuilder.scala [hunk 15]
```diff
@@ -203,6 +202,10 @@ class TaskBuilder(app: AppDefinition,
 
 object TaskBuilder {
 
+  sealed trait BuildResult
```
Comment (ConnorDoyle) [inline src/main/scala/mesosphere/mesos/TaskBuilder.scala]: How about adding `isSuccess` and `isFailure` methods to this trait?  It could clean up some of the use sites where you only care about which projection holds for your value.

File: src/test/scala/mesosphere/marathon/upgrade/AppStopActorTest.scala [hunk 16]
```diff
@@ -4,7 +4,7 @@ import akka.actor.{ ActorSystem, Props }
 import akka.testkit.{ TestProbe, TestActorRef, TestKit }
 import mesosphere.marathon.Protos.MarathonTask
 import mesosphere.marathon.event.{ HistoryActor, AppTerminatedEvent, MesosStatusUpdateEvent }
-import mesosphere.marathon.state.{ TaskFailure, TaskFailureRepository, AppDefinition, PathId }
+import mesosphere.marathon.state._
```
Comment (ConnorDoyle) [inline src/test/scala/mesosphere/marathon/upgrade/AppStopActorTest.scala]: Please revert this change.  Wildcard imports are to be avoided, especially for only four items.

File: src/test/scala/mesosphere/mesos/TaskBuilderTest.scala [hunk 17]
```diff
@@ -42,9 +42,10 @@ class TaskBuilderTest extends MarathonSpec {
       )
     )
 
-    assert(task.isDefined)
+    assert(task.isInstanceOf[TaskBuilder.BuildSuccess])
```
Comment (ConnorDoyle) [inline src/test/scala/mesosphere/mesos/TaskBuilderTest.scala]: This could read more nicely as `task.isSuccess`, see comments on `TaskBuilder.BuildResult` above.

File: src/test/scala/mesosphere/mesos/TaskBuilderTest.scala [hunk 18]
```diff
@@ -42,9 +42,10 @@ class TaskBuilderTest extends MarathonSpec {
...
 
-    val (taskInfo, taskPorts) = task.get
+    val success = task.asInstanceOf[TaskBuilder.BuildSuccess]
```
Comment (ConnorDoyle) [inline src/test/scala/mesosphere/mesos/TaskBuilderTest.scala]: Avoid `asInstanceOf`.  If you really want to cast and destructure unsafely, do it like this:

```scala
val TaskBuilder.BuildSuccess(taskInfo, taskPorts) = task.get
```

File: src/test/scala/mesosphere/mesos/TaskBuilderTest.scala [hunk 19]
```diff
@@ -105,9 +106,10 @@ class TaskBuilderTest extends MarathonSpec {
...
```
Comment (ConnorDoyle) [inline src/test/scala/mesosphere/mesos/TaskBuilderTest.scala]: No `asInstanceOf`.

**Comment by ConnorDoyle** [comment]
This PR addresses #937.

**Comment by ConnorDoyle** [comment]
@abhay-agarwal, this needs to be rebased on `master`.

**Comment by ConnorDoyle** [comment]
@abhay-agarwal, do you plan to finish this up?  We'd love to include this feature in 0.8.2 if possible.

**Comment by abhay-agarwal** [comment]
I hadn't looked at this since the coding challenge. I think I could do a better job, but it would be a context-switch that i'm not ready for pre Q1 release!

**Comment by drexin** [comment]
Closing this for now, feel free to update and re-open.
